File: math/gauss_legendre/erlang/gauss_legendre.erl
%% @author Mark Hillman <mark@markhillman.info>
%% @doc
%% This short module is used to calculate PI using the Gauss-Legendre
%% algorithm. The algorithm is fairly short and consists of just 4 steps.
%% The algorithm can be found on wikipedia [https://en.wikipedia.org/wiki/Gauss%E2%80%93Legendre_algorithm].

-module(gauss_legendre).
-export([calculate_pi/1]).

%% @doc
%% Initial values for the algorithm.
a_init() -> 1.
b_init() -> 1 / math:sqrt(2).
t_init() -> 1 / 4.
p_init() -> 1.

%% @doc
%% Calculate the next values of the iteration.
a_next(A, B) -> (A + B) / 2.
b_next(A, B) -> math:sqrt(A * B).
t_next(T, P, A, ANext) -> T - P * math:pow(A - ANext, 2).
p_next(P) -> 2 * P.

%% @doc
%% Determines current iteration of pi from the variables
pi_approx(A, B, T) ->
    math:pow(A + B, 2) / (4 * T).

%% @doc
%% Calculates PI using the Gauss-Legendre algorithm. The only argument is the
%% number of iterations that the algorithm should go through.
-spec calculate_pi(N :: pos_integer()) -> number().
calculate_pi(N) when N =< 0 -> 0;
calculate_pi(N) -> calculate_pi(N, a_init(), b_init(), t_init(), p_init(), 0).
calculate_pi(N, ANext, BNext, TNext, _P, N) -> pi_approx(ANext, BNext, TNext);
calculate_pi(N, A, B, T, P, Iter) ->
    ANext = a_next(A, B),
    BNext = b_next(A, B),
    TNext = t_next(T, P, A, ANext),
    PNext = p_next(P),
    calculate_pi(N, ANext, BNext, TNext, PNext, Iter+1).

File: math/gauss_legendre/python/gauss_legendre.py

'''
From Wikipedia:
https://en.wikipedia.org/wiki/Gauss%E2%80%93Legendre_algorithm

"The Gauss–Legendre algorithm is an algorithm to compute the digits of pi.
It is notable for being rapidly convergent, with only 25 iterations producing
45 million correct digits of π. However, the drawback is that it is computer memory-intensive"

'''

import numpy as np


num_iter = 5

#Instantiate
#The values at each iteration are stored
a = np.zeros(num_iter+1)
b = np.zeros(num_iter+1)
t = np.zeros(num_iter+1)
p = np.zeros(num_iter+1)
pi = np.zeros(num_iter+1)

a[0] = 1.0
b[0] = 1/np.sqrt(2)
t[0] = 0.25
p[0] = 1.0

#Update on each iteration
for i in range(1, num_iter+1):
  a[i] = (a[i-1]+b[i-1])/2
  b[i] = np.sqrt(a[i-1]*b[i-1])
  t[i] = t[i-1]-p[i-1]*(a[i-1]-a[i])**2
  p[i] = 2*p[i-1]
  
  pi[i] = (a[i]+b[i])**2/(4*t[i])
  
  print("{0:.20f}".format(pi[i]))